src/compiler/scala/tools/nsc/transform/CleanUp.scala

```diff
@@ -488,15 +488,34 @@ abstract class CleanUp extends Statics with Transform with ast.TreeDSL {
         super.transform(treeCopy.ArrayValue(rest, rest.elemtpt, elem0 :: rest.elems))
 
       // List(a, b, c) ~> new ::(a, new ::(b, new ::(c, Nil)))
-      case Apply(appMeth, List(Apply(wrapArrayMeth, List(StripCast(rest @ ArrayValue(elemtpt, _))))))
-      if wrapArrayMeth.symbol == currentRun.runDefinitions.Predef_wrapRefArray && appMeth.symbol == List_apply && rest.elems.length < transformListApplyLimit =>
+      // Seq(a, b, c) ~> new ::(a, new ::(b, new ::(c, Nil)))
+      case Apply(Select(appQual, nme.apply), List(Apply(wrapArrayMeth, List(StripCast(rest @ ArrayValue(elemtpt, _))))))
+        if wrapArrayMeth.symbol == currentRun.runDefinitions.Predef_wrapRefArray && {
+          val sym = appQual.symbol
+          sym == ListModule || sym == SeqModule || sym == ISeqModule
+        } && rest.elems.length < transformListApplyLimit =>
         val consed = rest.elems.reverse.foldLeft(gen.mkAttributedRef(NilModule): Tree)(
           (acc, elem) => New(ConsClass, elem, acc)
         )
         // Limiting extra stack frames consumed by generated code
         reducingTransformListApply(rest.elems.length) {
           super.transform(localTyper.typedPos(tree.pos)(consed))
         }
+      // Seq() ~> Nil
```

Comment:
`List() ~> Nil` is handled in the typer https://github.com/scala/scala/blob/9d4633dc5e3834154675152e6017ac708a9d2bf3/src/compiler/scala/tools/nsc/typechecker/Typers.scala#L3726-L3735

We should at put them all in the same place. To me, CleanUp makes sense. @retronym wdyt?

Reply:
IIRC the early optimization of `List()` => `Nil` was (at least historically) needed to enable some optimizations in the pattern matcher.

I think the new pattern matcher made some progress towards us _not_ needing that special case but I feel like it still wasn't straightforward to remove the special case.

https://github.com/scala/scala/blob/0bd07f9f8c24d88aba2daa29e5791d5204887f84/src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala#L404-L409
https://github.com/scala/scala/blob/0bd07f9f8c24d88aba2daa29e5791d5204887f84/src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala#L473